packages/terra-switch/src/Switch.jsx

```diff
@@ -0,0 +1,146 @@
+  OFF: 'Off',
+});
+
+const propTypes = {
+  /**
+    * Whether or not the Switch is enabled ("ON").
+    */
+  isOn: PropTypes.bool,
+
+  /**
+    * If true, the user won't be able to toggle the switch.
+    */
+  isDisabled: PropTypes.bool,
+
+  /**
+    * Text of the label.
+    */
+  labelText: PropTypes.node.isRequired,
+
+  /**
+    * Callback fired when the state is changed.
+    */
+  onChange: PropTypes.func,
+};
+
+const defaultProps = {
+  isOn: false,
+  isDisabled: false,
+  onChange: undefined,
+};
+
+const Switch = ({
+  isOn,
+  isDisabled,
+  onChange,
+  labelText,
+  ...customProps
+}) => {
+  const [focused, setFocus] = useState(false);
+  const sliderButtonRef = useRef(null);
+
+  const handleClick = (event) => {
+    setFocus(true);
+    sliderButtonRef.current.focus();
+    onChange(event);
+  };
+
+  const handleOnKeyDown = (event) => {
+    if (event.keyCode === KeyCode.KEY_SPACE || event.keyCode === KeyCode.KEY_RETURN) {
+      event.preventDefault();
+      setFocus(true);
+    }
+  };
+
+  const handleOnBlur = () => {
+    setFocus(false);
+  };
+
+  const handleKeyUp = (event) => {
+    if (event.keyCode === KeyCode.KEY_TAB) {
+      setFocus(true);
+    } else if (event.keyCode === KeyCode.KEY_SPACE || event.keyCode === KeyCode.KEY_RETURN) {
+      setFocus(true);
+      onChange(event);
+    }
+  };
+
+  const switchClassNames = cx([
+    'switch',
+    customProps.className,
+  ]);
+
+  const trayClassNames = cx([
+    'tray',
+    { selected: isOn },
+    { unselected: !isOn },
+    { 'is-disabled': isDisabled },
+  ]);
+
+  const sliderClassNames = cx([
+    'slider',
+    { 'is-disabled': isDisabled },
+    { 'is-focused': focused },
+  ]);
+
+  const statusLabelText = isOn ? SWITCH_STATE.ON : SWITCH_STATE.OFF;
+
+  const attributes = { ...customProps };
+  const switchButtonId = (attributes.id) ? attributes.id : `terra-switch-button-${uniqueid()}`;
+
+  return (
+    <div className={cx('switch-wrapper')}>
+      <label
+        htmlFor={switchButtonId}
+      >
+        <div className={cx('switch-container')}>
+          <div className={cx('label-container')}>
+            <span className={cx('label-text')}>{labelText}</span>
+            <span className={cx('status-label-text')}>{statusLabelText}</span>
+          </div>
+          <div
+            className={switchClassNames}
+          >
+            <span
+              className={trayClassNames}
+            >
+              <button
+                type="button"
+                id={switchButtonId}
+                disabled={isDisabled}
+                aria-checked={isOn}
+                aria-label={statusLabelText}
+                className={sliderClassNames}
+                role="switch"
+                tabIndex="0"
+                onClick={handleClick}
+                onKeyDown={handleOnKeyDown}
+                onKeyUp={handleKeyUp}
+                onBlur={handleOnBlur}
+                ref={sliderButtonRef}
+                {...customProps}
```

Comment:
Let's spread the custom props first. Here's a discussion around it - https://github.com/cerner/terra-core/issues/1248#issuecomment-365641007.

Reply:
moved to start : [here](https://github.com/cerner/terra-core/blob/1c002b1da07d995fc126d14539617437b6371dd7/packages/terra-switch/src/Switch.jsx#L119)